project-base/tests/ShopBundle/Functional/Model/Product/ProductVisibilityRecalculateConsumerTest.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\ShopBundle\Functional\Model\Product;
+
+use DateTime;
+use PhpAmqpLib\Message\AMQPMessage;
+use Shopsys\FrameworkBundle\Model\Product\Product;
+use Shopsys\FrameworkBundle\Model\Product\ProductDataFactoryInterface;
+use Shopsys\FrameworkBundle\Model\Product\ProductFacade;
+use Shopsys\FrameworkBundle\Model\Product\Visibility\ProductVisibilityRecalculateConsumerInterface;
+use Shopsys\ShopBundle\DataFixtures\Demo\ProductDataFixture;
+use Tests\ShopBundle\Test\TransactionFunctionalTestCase;
+
+class ProductVisibilityRecalculateConsumerTest extends TransactionFunctionalTestCase
+{
+    public function testConsumerChangesVisibilityOfProduct()
+    {
+        $productId = 4;
+
+        /** @var \Shopsys\FrameworkBundle\Model\Product\Product $product */
+        $product = $this->getReference(ProductDataFixture::PRODUCT_PREFIX . $productId);
+
+        /** @var \Shopsys\FrameworkBundle\Model\Product\ProductDataFactoryInterface $productDataFactory */
+        $productDataFactory = $this->getContainer()->get(ProductDataFactoryInterface::class);
+
+        /** @var \Shopsys\FrameworkBundle\Model\Product\ProductFacade $productFacade */
+        $productFacade = $this->getContainer()->get(ProductFacade::class);
+
+        /** @var \Shopsys\FrameworkBundle\Model\Product\Visibility\ProductVisibilityRecalculateConsumerInterface $productVisibilityRecalculateConsumer */
+        $productVisibilityRecalculateConsumer = $this->getContainer()->get(ProductVisibilityRecalculateConsumerInterface::class);
+
+        $productData = $productDataFactory->createFromProduct($product);
+        $productData->sellingTo = (new DateTime('-1 day'));
+
+        $productFacade->edit($product->getId(), $productData);
+
+        $message = new AMQPMessage($productId);
```

Comment:
I see that the acceptance criterion says "the consumer is tested automatically", however, what I am missing is an end to end test - why are you creating the message here directly? Should not it be handled by `$this->productQueueProducer->productChanged($product);` line in `ProductFacade::edit()`? Or it is not possible to test it this way?

Reply:
This is a bit problematic because of running consumers. Those run in different containers and we are not able to stop them in tests and start them again with tests configuration. That is why we set empty connection to rabbit in configuration and test it this way :)

Reply:
ok, fair enough :+1: Still, some user story for end to end tests of queues functionality would be nice, what do you say? :blush:

Reply:
I am not sure if that is possible

Reply:
https://media.giphy.com/media/CDIXdODZYXxD2/giphy.gif